### ISSUE TITLE
guvectorize run does not reproduce Python output

### ISSUE THREAD

**Issue by thoth291** [numba/numba#2364]
Here is the code which I'm trying to speedup:
```python
def resp_all(beta,dist):
    N,K = dist.shape
    min_dist = np.min(dist, axis = 1)
    max_dist = np.max(dist, axis = 1)
    dist_corr = (min_dist + max_dist) / 2
    dist_corr = np.minimum(dist_corr,(min_dist+700*(2/beta)))
    for n in range(0,N):
        dist[n,:] = dist[n,:] -  dist_corr[n]
        dist[n,:] = dist[n,:]* (-beta/2)      
    np.exp(dist, out=dist)
    Rsum = np.sum(dist,axis=1)
    for n in range(0,N):
        dist[n,:] = dist[n,:]/Rsum[n]
```

Python runtime is `944 ms`.
With `jit()` code runs in `749 ms`.

So I thought that I can do guvectorization of this code by unwrapping NumPy vectorization.
This is the result:
```python
def loop(dist,beta,min_dist,max_dist,dist_corr,res):
    #res is the same as dist
    K=dist.shape[0]
    min_dist[0]=dist[0]
    max_dist[0]=dist[0]
    for k in range(K):
        min_dist[0] = min(min_dist[0],dist[k])
        max_dist[0] = max(max_dist[0],dist[k])
    dist_corr[0]=min((min_dist[0] + max_dist[0]) / 2.0,(min_dist[0]+700*(2/beta[0])))
    max_dist[0]=0.0
    for k in range(K):
        res[k] = np.exp(-beta[0]*(dist[k] -  dist_corr[0])/2.0)
        max_dist[0]=max_dist[0]+dist[k]
    for k in range(K):
        res[k] = res[k]/max_dist[0]


guloop=guvectorize(['void(float64[:], float64[:],float64[:],float64[:], float64[:],float64[:])'],
                    '(n),(),(),(),()->(n)', target='parallel',nopython=True)(loop)

def resp_all_dummy2(beta,min_dist,max_dist,dist_corr,dist):
    N,K = dist.shape
    for n in range(N):
        loop(dist[n],[beta[n]],[min_dist[n]],[max_dist[n]],[dist_corr[n]],dist[n])
```

In pure Python this code `resp_all_dummy2` runs in `3 min 37 sec` and producing proper output.
But when I run guvectorized code
```python
res=guloop(dist5,beta,min_dist,max_dist,dist_corr)
```
it runs faster `355 ms` but producing wrong output.

**Could you please help me to figure out what is going on?**
Here is the [notebook](https://anaconda.org/thoth/resp_all/notebook) which explains this issue in details.

Another question is that: **if there is any way to utilize parallel guvectorization for this code `without` unwrapping NumPy vectorization?**

**Comment by ninegua** [comment]
In your `resp_all_dummy2` function, the call to `loop` passes the value of `dist[n]` to both of the `dist` and `res` parameters of the loop function. This makes `dist` and `res` alias each other. However, when you `guvectorize` the loop function and call it without having to pass an array to `res`, the `res` array is created afresh. So it will not alias `dist`, and hence the difference.

I haven't tried this but think the output will be correct if you change:

````
        res[k] = np.exp(-beta[0]*(dist[k] -  dist_corr[0])/2.0)
        max_dist[0]=max_dist[0]+dist[k]
```

to 

```
        res[k] = np.exp(-beta[0]*(dist[k] -  dist_corr[0])/2.0)
        max_dist[0]=max_dist[0]+res[k]
```

**Comment by ninegua** [comment]
BTW, I think `min_dist,max_dist,dist_corr` can be kept as local variables instead of parameters.

**Comment by thoth291** [comment]
Thank you, @ninegua .

Good catch! I updated the code `dist` to `rest` at the `max_dist` line - but I **still get the wrong output**.

I'm afraid to ask - but could you please make a closer look - may be my signatures are wrong...

You've said:
> `min_dist,max_dist,dist_corr` can be kept as local variables

What do you mean? How should I do that?

Please excuse me and let me repeat myself: **if there is any way to utilize parallel guvectorization for this code without unwrapping NumPy vectorization?** - can this be a path of help in this current situation?

**Comment by ninegua** [comment]
I spotted another glitch in your "de-vectorized" loop function. The original code says `dist[n,:] = dist[n,:]* (-beta/2)`, which means each row of dist times each row of beta. But what your loop function does is actually equivalent to `dist[n] = dist[n] * (-beta[n]/2)`. This should explain the output difference. So which one do you actually mean?

In any case, the result I get from `guloop` is same as the `resp_all_dummy2`, and both are different than the result from `resp_all`.

As for your last question, we are actually working on an automatic parallelization feature. Some initial progress is in [this pull request](https://github.com/numba/numba/pull/2318). Although it only supports basic reduction and array operations, hopefully given more time we'll be able to cover reduction over a given axis, as well as range selection. Then your original `resp_all` should be able to run in parallel with no modification.

**Comment by ninegua** [comment]
Here is a slightly simplified version of `loop` that follows `dist[n] = dist[n] * (-beta[n]/2)` if that is what you meant:

```
def loop(dist,beta,res):
    #res is the same as dist
    K=dist.shape[0]
    min_dist=dist[0]
    max_dist=dist[0]
    for k in range(K):
        min_dist = min(min_dist,dist[k])
        max_dist = max(max_dist,dist[k])
    dist_corr=min((min_dist + max_dist) / 2.0,(min_dist+700*(2/beta)))
    max_dist=0.0
    for k in range(K):
        res[k] = np.exp(-beta*(dist[k] -  dist_corr)/2.0)
        max_dist=max_dist+res[k]
    for k in range(K):
        res[k] = res[k]/max_dist

guloop=guvectorize(['void(float64[:],float64,float64[:])'],
                    '(n),()->(n)', target='parallel',nopython=True)(loop)

dist5=origdist.copy()
res=guloop(dist5,beta)
```

If instead what you really meant was `dist[n] = dist[n] * (-beta/2)`, then:

```
def loop(dist,beta0,beta,res):
    #res is the same as dist
    K=dist.shape[0]
    min_dist=dist[0]
    max_dist=dist[0]
    for k in range(K):
        min_dist = min(min_dist,dist[k])
        max_dist = max(max_dist,dist[k])
    dist_corr=min((min_dist + max_dist) / 2.0,(min_dist+700*(2/beta0)))
    max_dist=0.0
    for k in range(K):
        res[k] = np.exp(-beta[k]*(dist[k] -  dist_corr)/2.0)
        max_dist=max_dist+res[k]
    for k in range(K):
        res[k] = res[k]/max_dist

guloop=guvectorize(['void(float64[:],float64,float64[:],float64[:])'],
                    '(n),(),(k)->(n)', target='parallel',nopython=True)(loop)

dist5=origdist.copy()
res=guloop(dist5,beta,beta)
```

**Comment by thoth291** [comment]
Ah, I see what you mean!
I have to confess - this is not my code.
So I just tried to speed it up without looking into other parts of the program.
My apologies for that. I'll be careful in the future!

Essentially this line:
```python
dist_corr = np.minimum(dist_corr,(min_dist+700*(2/beta)))
```
And this line:
```python
dist[n,:] = dist[n,:]* (-beta/2)
```
Are in conflict if one assumes that **beta** is a vector (as I did), but it's not a vector - **in fact it's a constant**.

I have tried to separate vectorization part of the loop one more time - but it didn't work even for Python - so I guess I'm just doing something very stupid.

Thank you for your help.
Unwrapping the loops back from NumPy is definitely a pain - and I hope that sometimes in the future Numba will be smart enough to help with that just as you mentioned.
